src/NHibernate/Engine/BatchFetchQueue.cs

```diff
@@ -143,23 +187,42 @@ public object[] GetCollectionBatch(ICollectionPersister collectionPersister, obj
 			int end = -1;
 			bool checkForEnd = false;
 
-			// this only works because collection entries are kept in a sequenced
-			// map by persistence context (maybe we should do like entities and
-			// keep a separate sequences set...)
-			foreach (DictionaryEntry me in context.CollectionEntries)
+			if (batchLoadableCollections.TryGetValue(collectionPersister.Role, out var map))
 			{
-				CollectionEntry ce = (CollectionEntry) me.Value;
-				IPersistentCollection collection = (IPersistentCollection) me.Key;
-				if (!collection.WasInitialized && ce.LoadedPersister == collectionPersister)
+				foreach (KeyValuePair<CollectionEntry,IPersistentCollection> me in map)
 				{
+					var ce = me.Key;
+					var collection = me.Value;
+					if (ce.LoadedKey == null)
+					{
+						// the LoadedKey of the CollectionEntry might be null as it might have been reset to null
+						// (see for example Collections.ProcessDereferencedCollection()
+						// and CollectionEntry.AfterAction())
+						// though we clear the queue on flush, it seems like a good idea to guard
+						// against potentially null LoadedKey:s
+						continue;
+					}
+
+					if (collection.WasInitialized)
+					{
+						// should never happen
+						if (log.IsWarnEnabled())
+						{
+							log.Warn("Encountered initialized collection in BatchFetchQueue, this should not happen.");
+						}
```

Comment:
I personally tend to minimize such code by eliminating the "isEnabled" guard for simple logs, since the logger itself is supposed to test this. I keep the guard only if the log involve more than calling the logger, like having some string computing before calling it.
The "should never happen" comment is also a bit redundant with the log message. So I would tend to only have one line here instead of five, the one doing the log.